### PR TITLE
Implement monster attack style immunities and resistances

### LINKED ISSUE #20: calc: monster immunities
### Melee
- [ ] Kraken
- [ ] Abyssal portal
- [ ] TzKal-Zuk
- [ ] Jal-MejJak
- [ ] The Leviathan
- [ ] Avianses, except when using a salamander
- [ ] Dawn
- [ ] Olm mage hand and head (immune, not 1/3 dmg)
- [ ] Zulrah

### Ranged
- [ ] Tekton
- [ ] Kraken (1/7 dmg, min 1)
- [ ] Dusk
- [ ] Olm hands (1/3 dmg)
- [ ] Glowing crystal

### Magic
- [ ] Dusk
- [ ] Tekton (1/5 dmg, min 1)
- [ ] Glowing crystal (1/3 dmg)
- [ ] Olm head and melee hand (1/3 dmg)

### Weapon-Specific
- [ ] Guardians
- [ ] T3 Vampyres
    - Requires one of Ivandis' flail, Blisterwood sickle, or Blisterwood flail

### PR DESCRIPTION
**Pull request by Nightfirecat**
Fixes #20

### CODE REVIEW

File: src/lib/CombatCalc.ts [hunk 1]
```diff
@@ -922,35 +1012,62 @@ export default class CombatCalc {
...
     return dist;
   }
-  
+
   isImmune(): boolean {
-    // todo
-    return false;
+    const monsterId = this.monster.id || 0;
+    const styleType = this.player.style.type;
+    const meleeStyles:CombatStyleType[] = [ 'crush', 'slash', 'stab' ];
```
Comment (jayktaylor) [inline src/lib/CombatCalc.ts:1056]: You can use `this.isUsingMeleeStyle()` here instead

File: src/lib/CombatCalc.ts [hunk 2]
```diff
@@ -247,11 +264,11 @@ export default class CombatCalc {
 
   private isWearingSilverWeapon(): boolean {
     return this.wearing([
```
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:266]: This is my bad for not doing it originally, but can you think of a good way to add silver bolt detection here? Similar to below for leaf-bladed I'm fine with not checking that we're using an applicable weapon for now.

File: src/lib/CombatCalc.ts [hunk 3]
```diff
@@ -266,6 +283,79 @@ export default class CombatCalc {
...
+   * @see https://oldschool.runescape.wiki/w/Corpbane_weapons
+   * @todo Handle ruby bolt procs separately (non-procs do half damage, but procs deal full damage)
+   */
+  private isWearingCorpbaneWeapon(): boolean {
+    const { weapon } = this.player.equipment;
+    if (!weapon) {
+      return false;
+    }
+
+    if (weapon.name === 'Osmumten\'s fang') {
```
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:340]: Doesn't account for (or), but we also have `isWearingFang`

File: src/lib/CombatCalc.ts [hunk 4]
```diff
@@ -922,35 +1012,62 @@ export default class CombatCalc {
...
+    const monsterId = this.monster.id || 0;
+    const styleType = this.player.style.type;
+    const meleeStyles:CombatStyleType[] = [ 'crush', 'slash', 'stab' ];
+    return (IMMUNE_TO_MAGIC_DAMAGE_NPC_IDS.includes(monsterId) && styleType === 'magic')
+      || (IMMUNE_TO_RANGED_DAMAGE_NPC_IDS.includes(monsterId) && styleType === 'ranged')
+      || (IMMUNE_TO_MELEE_DAMAGE_NPC_IDS.includes(monsterId) && meleeStyles.includes(styleType))
+      || (IMMUNE_TO_NON_SALAMANDER_MELEE_DAMAGE_NPC_IDs.includes(monsterId) && meleeStyles.includes(styleType) && this.player.equipment.weapon?.category !== EquipmentCategory.SALAMANDER)
+      || (this.monster.attributes.includes(MonsterAttribute.VAMPYRE_3) && !this.isWearingIvandisWeapon())
+      || (GUARDIAN_IDS.includes(monsterId) && this.player.equipment.weapon?.category !== EquipmentCategory.PICKAXE)
+      || (this.monster.attributes.includes(MonsterAttribute.LEAFY) && !this.isWearingLeafBladedWeapon());
```
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:1063]: this is hard to read, can you expand it to a sequence of if-checks instead?

**Review by LlemonDuck — CHANGES_REQUESTED** [review]

File: src/lib/CombatCalc.ts (hunk 2 above)
Comment (Nightfirecat) [reply src/lib/CombatCalc.ts:271]: I've added a check for silver bolts + ranged style before this broader gear check block.

**Review by LlemonDuck — APPROVED** [review]

**Pull request merged**